src/plot.cpp

```diff
@@ -1582,6 +1532,273 @@ void ProjectionPlot::set_field_of_view(pugi::xml_node node)
   }
 }
 
+PhongPlot::PhongPlot(pugi::xml_node node) : RayTracePlot(node)
+{
+  set_opaque_ids(node);
+  set_diffuse_fraction(node);
+  set_light_position(node);
+}
+
+void PhongPlot::print_info() const
+{
+  fmt::print("Plot Type: Phong\n");
+  RayTracePlot::print_info();
+}
+
+void PhongPlot::create_output() const
+{
+  size_t width = pixels_[0];
+  size_t height = pixels_[1];
+  ImageData data({width, height}, not_found_);
+
+#pragma omp parallel
+  {
+
+#ifdef _OPENMP
+    const int n_threads = omp_get_max_threads();
+    const int tid = omp_get_thread_num();
+#else
+    int n_threads = 1;
+    int tid = 0;
+#endif
+
+    int vert = tid;
+    for (int iter = 0; iter <= pixels_[1] / n_threads; iter++) {
+      if (vert < pixels_[1]) {
+        for (int horiz = 0; horiz < pixels_[0]; ++horiz) {
+
+          // RayTracePlot implements camera ray generation
+          std::pair<Position, Direction> ru = get_pixel_ray(horiz, vert);
+          PhongRay ray(ru.first, ru.second, *this);
+
+          ray.trace();
+          data(horiz, vert) = ray.result_color();
+        }
+      } // end "if" vert in correct range
+
+      // TODO: can maybe remove this barrier
+#pragma omp barrier
+      vert += n_threads;
+    }
+  } // end omp parallel
+
+#ifdef USE_LIBPNG
+  output_png(path_plot(), data);
+#else
+  output_ppm(path_plot(), data);
+#endif
+}
+
+void PhongPlot::set_opaque_ids(pugi::xml_node node)
+{
+  if (check_for_node(node, "opaque_ids")) {
+    opaque_ids_ = get_node_array<int>(node, "opaque_ids");
+    // It is read in as actual ID values, but we have to convert to indices in
+    // mat/cell array
+    for (auto& x : opaque_ids_)
+      x = color_by_ == PlotColorBy::mats ? model::material_map[x]
+                                         : model::cell_map[x];
+  }
+  // We make sure the list is sorted in order to later use
+  // std::binary_search.
+  std::sort(opaque_ids_.begin(), opaque_ids_.end());
+}
+
+bool PhongPlot::is_id_opaque(int id) const
+{
+  return std::binary_search(opaque_ids_.begin(), opaque_ids_.end(), id);
```

Comment:
If we use a `std::set` we can avoid the search.

Reply:
I thought about using a `std::set`, but we don't use it anywhere else AFAIK so it's nice to leave out that big ol' header. Also, under the hood it just is using a tree to keep data sorted, and checking for presence of an element would look a lot like binary search. For this reason I opted for an `std::vector` we sort one time after reading input.

Reply:
Set is doing something equivalent to a binary search under the hood, so I'm resolving this one.

Reply:
Do me a favor and put a TODO comment in here to try `std::unordered_set` to test its performance? 

I typed `std::set` above and you're right that the value containment check is O(log N) for `std::set`. But using `std::unordered_set` really should be faster as it's using a hash table lookup, but if this isn't a performance bottleneck for us and we want to avoid the header then I get the tradeoff. 

It'd be nice to have a note in here though as a flagpost for someone looking to improve performance in the future.

Reply:
Yessir! Sorry, probably should not have resolved that without you OK'ing it. 😄

Reply:
Did that comment get added? Maybe your local branch 
needs a push?